x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/UserRoleMapper.java

```diff
@@ -145,6 +148,65 @@ public Map<String, Object> getMetadata() {
         public RealmConfig getRealm() {
             return realm;
         }
+
+        // Package private for testing
+        DistinguishedNameNormalizer getDnNormalizer() {
+            return new DistinguishedNameNormalizer();
+        }
+    }
+
+    /**
+     * This class parse the given string into a DN and return its normalized format.
+     * If the input string is not a valid DN, {@code null} is returned.
+     * The DN parsing and normalization are cached internally so that the same
+     * input string will only be processed once (as long as the cache entry is not GC'd).
+     * The cache works regardless of whether the input string is a valid DN.
+     *
+     * The cache uses {@link SoftReference} for its values so that they free for GC.
+     * This is to prevent potential memory pressure when there are many concurrent role
+     * mapping processes coupled with large number of groups and role mappings, which
+     * in theory is unbounded.
+     */
+    class DistinguishedNameNormalizer {
+        private static final Logger LOGGER = LogManager.getLogger(DistinguishedNameNormalizer.class);
+        private static final SoftReference<String> NULL_REF = new SoftReference<>(null);
+        private final Map<String, SoftReference<String>> cache = new HashMap<>();
```

Comment:
Since the cahce now stores String, one alternative to `SoftReference` is to use `String.intern()` which can potentially have even better memory footprint when there are large number of duplicates (e.g. high number of concurrent fresh authentication). But I find it is not straightforward to reason about the lifecycle of interned string. It also has minor performance implications due to deduplication. The scenario it is really great for is not all that common as well. Overall, I feel SoftReference is still the better choice.